vllm/v1/cudagraph_dispatcher.py

```diff
@@ -0,0 +1,121 @@
+        create_mixed_batch_full_cg = attn_cg_support in [
+                            AttentionCGSupport.ALWAYS_UNIFIED,
+                            AttentionCGSupport.ALWAYS_SEPARATE] and \
+                            capture_mixed_batches
+        if self.cudagraph_mode == CUDAGraphMode.FULL:
+            # full cudagraph for mix prefill-decode/general batches
+            if create_mixed_batch_full_cg:
+                for bs in self.compilation_config.cudagraph_capture_sizes:
+                    self.add_cudagraph_key(
+                        CUDAGraphMode.FULL,
+                        BatchDescriptor(num_tokens=bs, is_uniform=False))
+
+            # always create full cudagraph for uniform batches if cudagraph
+            # separate routine is enabled.
+            if self.compilation_config.cudagraph_separate_routine:
+                max_num_tokens = uniform_decode_query_len * \
+                    self.vllm_config.scheduler_config.max_num_seqs
+                cudagraph_capture_sizes_for_decode = [
+                    x for x in self.compilation_config.cudagraph_capture_sizes
+                    if x <= max_num_tokens and x >= uniform_decode_query_len
+                ]
+                for bs in cudagraph_capture_sizes_for_decode:
+                    self.add_cudagraph_key(
+                        CUDAGraphMode.FULL,
+                        BatchDescriptor(num_tokens=bs, is_uniform=True))
+        self.keys_initialized = True
+
+    def dispatch(
+        self, batch_descriptor: BatchDescriptor
+    ) -> tuple[CUDAGraphMode, Optional[BatchDescriptor]]:
+        # if not initialized, just skip dispatching.
+        if not self.keys_initialized:
+            logger.warning_once("cudagraph dispatching keys are not "
+                                "initialized. No cudagraph will be used.")
+            return CUDAGraphMode.NONE, None
+
+        assert batch_descriptor.is_uniform is not None, \
+            "is_uniform is required for cudagraph dispatching."
+
+        # always treat is_uniform=False if cudagraph_separate_routine disabled.
+        if not self.compilation_config.cudagraph_separate_routine:
+            batch_descriptor = BatchDescriptor(
+                num_tokens=batch_descriptor.num_tokens, is_uniform=False)
+
+        # check if key exists for full cudagraph
+        if batch_descriptor in self.cudagraph_keys[CUDAGraphMode.FULL]:
+            return CUDAGraphMode.FULL, batch_descriptor
+
+        # otherwise, check if key exists for more "general" piecewise cudagraph
+        new_batch_descriptor = BatchDescriptor(
+            num_tokens=batch_descriptor.num_tokens, is_uniform=None)
+        if new_batch_descriptor in self.cudagraph_keys[
+                CUDAGraphMode.PIECEWISE]:
+            return CUDAGraphMode.PIECEWISE, new_batch_descriptor
+
+        # finally, just return no cudagraphs
+        return CUDAGraphMode.NONE, None
```

Comment:
I think `dispatch` should only care about keys that exist and not look at config at all, which would simplify the logic, especially after we get rid of `None`:

```suggestion
        # Check if key exists for full cudagraph
        if batch_descriptor in self.cudagraph_keys[CUDAGraphMode.FULL]:
            return CUDAGraphMode.FULL, batch_descriptor

        # CUDA Graphs captured for non-uniform batches should also support uniform ones
        if batch_descriptor.non_uniform() in self.cudagraph_keys[CUDAGraphMode.FULL]:
            return CUDAGraphMode.FULL, batch_descriptor.non_uniform()

        # Otherwise, check if key exists for piecewise cudagraph
        if batch_descriptor in self.cudagraph_keys[
                CUDAGraphMode.PIECEWISE]:
            return CUDAGraphMode.PIECEWISE, new_batch_descriptor
            
        # For piecewise, check non-uniform as well
        if batch_descriptor.non_uniform() in self.cudagraph_keys[
                CUDAGraphMode.PIECEWISE]:
            return CUDAGraphMode.PIECEWISE, batch_descriptor.non_uniform()

        # finally, just return no cudagraphs
        return CUDAGraphMode.NONE, None
```